File: baekjoon_all/01000+/boj_1654.cpp
// Solve 2022-07-24
// Update 2025-02-21

#include <bits/stdc++.h>

#define FASTIO ios_base::sync_with_stdio(false);cin.tie(NULL);
#define SIZE(v) (int)v.size()
#define ALL(v) v.begin(),v.end()
#define SETW(n, c) cout << setw(n) << setfill(c);
#define SETP(n) cout << fixed << setprecision(n);

using namespace std;
using ll = long long;
using uint = unsigned int;
using ull = unsigned long long;
using ld = long double;
using pii = pair<int, int>;

int lengths[10000];

int main() {
    FASTIO;

    int k, n;
    cin >> k >> n;

    for (int i = 0; i < k; i++) {
        cin >> lengths[i];
    }

    ll low = 1;
    ll high = 2147483648LL;

    while (low + 1 < high) {
        ll mid = (low + high) >> 1;
        ll mid_val = 0;

        for (int i = 0; i < k; i++) {
            mid_val += lengths[i] / mid;
        }

        if (mid_val >= n) {
            low = mid;
        }
        else {
            high = mid;
        }
    }

    cout << low << '\n';

    return 0;
}

File: baekjoon_all/01000+/boj_1654.java
// Solve 2024-01-28
// Update 2025-02-21

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class boj_1654 {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");
        int k = Integer.parseInt(st.nextToken());
        int n = Integer.parseInt(st.nextToken());
        int[] lengths = new int[k];

        for (int i = 0; i < k; i++) {
            lengths[i] = Integer.parseInt(br.readLine());
        }

        long low = 1L;
        long high = (1L << 31);

        while (low + 1 < high) {
            long mid = (low + high) / 2;
            long val = 0L;

            for (int length : lengths) {
                val += length / mid;
            }

            if (val >= n) {
                low = mid;
            } else {
                high = mid;
            }
        }

        System.out.println(low);
    }

}
